platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,762 @@
+                        first: index == 0,
+                        last: false,
+                        buff_size: 0x1000,
+                        buff_addr: unsafe { buf_ptr.add(index * 0x1000) },
+                        next_desc: Some(&descriptors[index + 1] as *const _ as _),
+                    })
+                    .map_err(|_| sdmmc::Error::Other)?;
+                    remaining -= 0x1000;
+                    index += 1;
+                }
+
+                descriptors[index] = idmac::Descriptor::try_from(idmac::DescriptorConfig {
+                    disable_int_on_complete: false,
+                    first: index == 0,
+                    last: true,
+                    buff_size: remaining as u16,
+                    buff_addr: unsafe { buf_ptr.add(index * 0x1000) },
+                    next_desc: None,
+                })
+                .map_err(|_| sdmmc::Error::Other)?;
+
+                // TODO: should this return some kind of guard, which needs to be used later
+                // to make sure the descriptor has a long enough lifetime?
+                unsafe { self.prepare_dma(&descriptors[0], *cnt as u32) };
+            }
+            _ => (),
+        }
+
+        // Set the argument to be sent
+        self.smhc
+            .smhc_cmdarg
+            .write(|w| unsafe { w.bits(params.argument) });
+
+        // Configure and start command
+        self.smhc.smhc_cmd.write(|w| {
+            w.cmd_load().set_bit();
+            w.wait_pre_over().wait();
+            w.stop_cmd_flag().bit(auto_stop);
+            w.data_trans().bit(data_trans);
+            w.trans_dir().bit(trans_dir);
+            w.chk_resp_crc().bit(chk_resp_crc);
+            w.long_resp().bit(long_resp);
+            w.resp_rcv().bit(resp_rcv);
+            w.cmd_idx().variant(cmd_idx);
+            w
+        });
+
+        guard.data.state = State::WaitForCommand;
+        // Now wait for completion or error interrupt
+        guard.wait_for_irq().await;
+        tracing::trace!("SMHC operation completed");
+        let res = if let Some(error) = guard.data.err.take() {
+            tracing::warn!(?error, "SMHC error");
+            Err(sdmmc::Error::Other) // TODO
+        } else {
+            if long_resp {
+                Ok(sdmmc::Response::Long([
+                    self.smhc.smhc_resp0.read().bits(),
+                    self.smhc.smhc_resp1.read().bits(),
+                    self.smhc.smhc_resp2.read().bits(),
+                    self.smhc.smhc_resp3.read().bits(),
+                ]))
+            } else {
+                Ok(sdmmc::Response::Short {
+                    value: self.smhc.smhc_resp0.read().bits(),
+                    data: match core::mem::replace(&mut guard.data.op, SmhcOp::None) {
+                        SmhcOp::Read { buf, .. } => Some(buf),
+                        _ => None,
+                    },
+                })
+            }
+        };
+        res
```

Comment:
style nit: i think this can be simplified a bit to just:
```suggestion
        if let Some(error) = guard.data.err.take() {
            tracing::warn!(?error, "SMHC error");
            Err(sdmmc::Error::Other) // TODO
        } else if long_resp {
            Ok(sdmmc::Response::Long([
                self.smhc.smhc_resp0.read().bits(),
                self.smhc.smhc_resp1.read().bits(),
                self.smhc.smhc_resp2.read().bits(),
                self.smhc.smhc_resp3.read().bits(),
            ]))
        } else {
            Ok(sdmmc::Response::Short {
                value: self.smhc.smhc_resp0.read().bits(),
                data: match core::mem::replace(&mut guard.data.op, SmhcOp::None) {
                    SmhcOp::Read { buf, .. } => Some(buf),
                    _ => None,
                },
            })
        }
```